Added mutex lock to prevent race condition in accessing `m.returnValues` in `Times` condition

me-nutant opened the pull request:
### Issue
Found a race condition while using this framework where multiple go routines are entering critical section in `generateReplacement` function causing panic with error `out of index`.

#### Example:
- Consider a function `Foo()` and its corresponding test function `TestFoo()`.
- The logic in `Foo` contains spawning of `go routines` and each of them are calling a function `Bar()`.
- In our `TestFoo`, we set the state in such a way that `Bar` will be called exactly 2 times. So, we mock the function `Bar` with `Times(2)` setting `m.returnValues[0].times=2` and `Return` some values which appends a single element in `m.returnValues`.
- Our critical section start from the line (say L1): `m.returnValues[0].calledTimes++` and ends after the `return` from this function.
- In some case, it is possible that both the go routines reached `L1` which makes the value of `m.returnValues[0].calledTimes` equalling `2`. SO, `m.returnValues[0].calledTimes` equals `m.returnValues[0].times`.
- In this case, it is possible only one of the go routine reaches the line `m.returnValues = m.returnValues[1:]` expelling the only element in `m.returnValues`.
- Later, the second go routine reaches the line `rets := m.returnValues[0]`. Here, it is trying to access the first element of an empty slice `m.returnValues` causing panic with error `out of index`.

### Resolution
Using `mutex lock` in such a way that only one go routine can access `m.returnValues` and therefore, only one go routine can increase `m.returnValues[0].calledTimes`.

me-nutant committed:
Added mutex lock to prevent race condition in accessing m.returnValues in `Times` condition

cheerioskun commented:
What is the state you are trying to mutually exclude? (which variables)
will the conditional check before the lock not be invalidated for the runner who lost the lock. 

ex. T1 and T2 arrive at condition and enter the if block. Then T1 took the lock. When it released the lock to T2, the condition with which T2 entered this thread may not be true anymore (calledTimes < times)

me-nutant committed:
Changes:
	- moved the lock and updating the `calledTimes` above the condition so that more than expected calls
		will raise the panic of unexpected call to that function

me-nutant commented:
> ex. T1 and T2 arrive at condition and enter the if block. Then T1 took the lock. When it released the lock to T2, the condition with which T2 entered this thread may not be true anymore (calledTimes < times)

@cheerioskun, I have moved the taking of lock and updating the `calledTimes` above the check of `calledTimes < times` to prevent this case. Now, the go routines can't enter the condition before increasing the `calledTimes` and more than expected calls will raise a `panic` according the code logic.

cheerioskun reviewed (approved):
i don't know if I have write access. Please also try running all of our test suite with this patch

me-nutant commented:
> i don't know if I have write access. Please also try running all of our test suite with this patch

It seems you don't have the write access. The merge is still blocked
